### PR TITLE
SD-2520: (sd-2666) - UI: Matching/Unmatching stories from publish filter

### PR DESCRIPTION
**Pull request by sjunaid**

### CODE REVIEW

**Comment by akintolga** [comment]
@sjunaid Would it make sense to separate the test logic against its own controller? The existing controller has a test() function and now there's a test_filter() function, it is getting a a bit confusing.

**Comment by akintolga** [comment]
:+1:

**Comment by sivakuna-aap** [comment]
:+1: Could have been better with 2 commits. You can amend the commits until the commit isn't submitted for review.

File: client/app/scripts/superdesk-publish/filters/publish-filters.js [hunk 1]
```diff
@@ -168,8 +172,8 @@ function FilterConditionsController($scope, filters, notify, modal) {
 
 }
 
-PublishFiltersController.$inject = ['$scope', 'filters', 'notify', 'modal'];
-function PublishFiltersController($scope, filters, notify, modal) {
+PublishFiltersController.$inject = ['$scope', 'filters', 'notify', 'modal', '$rootScope'];
+function PublishFiltersController($scope, filters, notify, modal, $rootScope) {
```
Comment (plamut) [inline client/app/scripts/superdesk-publish/filters/publish-filters.js]: `$rootScope` is not used anywhere in this controller?

File: client/app/scripts/superdesk-publish/filters/publish-filters.js [hunk 2]
```diff
@@ -265,6 +269,10 @@ function PublishFiltersController($scope, filters, notify, modal) {
         $scope.previewPublishFilter();
     };
 
+    $scope.productionTest = function (filter) {
+        $scope.$broadcast('triggerTest', filter);
```
Comment (plamut) [inline client/app/scripts/superdesk-publish/filters/publish-filters.js]: ...or did you mean to say `$rootScope.$broadcast(...)` here?

File: client/app/scripts/superdesk-publish/filters/publish-filters.js [hunk 3]
```diff
@@ -352,8 +360,116 @@ function PublishFiltersController($scope, filters, notify, modal) {
...
+function ProductionTestController($scope, filters, notify, modal, $location, $window, $rootScope) {
+    $scope.preview = null;
+    $scope.selected = {};
+    $scope.selectedItem = {};
+    var UP = -1,
+    DOWN = 1,
+    MOVES = {
+        38: UP,
+        40: DOWN
+    };
```
Comment (plamut) [inline client/app/scripts/superdesk-publish/filters/publish-filters.js]: `Object.freeze()` could be used here, but nothing significant.

File: client/app/scripts/superdesk-publish/filters/publish-filters.js [hunk 4]
```diff
@@ -352,8 +360,116 @@ function PublishFiltersController($scope, filters, notify, modal) {
...
+            event.stopImmediatePropagation();
+        }
+    }
+    $scope.fetchResults = function() {
+        fetchProductionTestResult();
+    };
+
+    $scope.openViewTab = function(guid) {
+        var _url = '#/authoring/' + guid + '/view';
+        $window.open(_url, '_blank');
```
Comment (plamut) [inline client/app/scripts/superdesk-publish/filters/publish-filters.js]: Styling tip: no real need to use the underscore prefix here, it's obvious that `url` is local to the function.

File: client/app/scripts/superdesk-publish/filters/view/production-test-view.html [hunk 5]
```diff
@@ -0,0 +1,51 @@
+<section class="main-section.filter">
+    <div class="preview-layout" ng-class="{closed: !selected.preview}">
+        <div class="list-pane">
+            <header>
+                <div class="sortbar pull-left">
+                    <span ng-repeat="type in resultType">
+                        <input type="radio" ng-click="fetchResults()" name="type" value="{{type.value}}" data-ng-model="model.selectedType"> {{type.id}}
```
Comment (plamut) [inline client/app/scripts/superdesk-publish/filters/view/production-test-view.html]: `type.value` does probably not change, and we can thus use one-time binding for it?

File: client/app/scripts/superdesk-publish/filters/view/production-test-view.html [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+                                <th translate>Unique Name</th>
+                                <th translate>Headline</th>
+                                <th translate>Keywords</th>
+                                <th translate>Source</th>
+                                <th translate>ANPA Category</th>
+                                <th translate>View</th>
+                            </tr>
+                        </thead>
+                        <tbody>
+                            <tr ng-class="{active: result_item === selectedItem, archived: result_item.archived || result_item.created, locked: lock.isLocked}" tabindex="1" ng-keyup="handleKeyEvent($event)" ng-repeat="result_item in testResult track by result_item._id" ng-click="preview(result_item);">
```
Comment (plamut) [inline client/app/scripts/superdesk-publish/filters/view/production-test-view.html]: This line is really long and difficult to comprehend, it requires a lot of horizontal scrolling. I suggest splitting it into multiple lines for better readability.

File: client/app/scripts/superdesk-publish/filters/view/production-test-view.html [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+                                <td><i class="filetype-icon-{{result_item.type}}" ng-hide="hover || item.selected"></i></td>
+                                <td><div class="state-label state-{{result_item.state}}">{{ result_item.state}}</div></td>
+                                <td>{{ result_item.unique_name}}</td>
+                                <td>{{ result_item.headline | limitTo: 60}}
+                                    <span sd-item-container data-item="result_item" title="{{ result_item.container }}" 
+                                    class="container"></span>
+                                </td>
+                                <td>{{ result_item.slugline | limitTo: 40}}</td>
+                                <td>{{ result_item.source}}</td>
+                                <td>{{ result_item['anpa-category'].name }}</td>
```
Comment (plamut) [inline client/app/scripts/superdesk-publish/filters/view/production-test-view.html]: Please check if any of these result_item's properties do not change and could be one-time bound, reducing the number of `$watches`.

**Comment by plamut** [comment]
Tests are still missing, and there's quite a lot of new code, thus we should really add them.

**Comment by plamut** [comment]
:+1:

Looks good now after the cleanup , although I haven't checked the test coverage for `ProductionTestController` (I assume it's sufficient?).

**Comment by akintolga** [comment]
:+1:

**Pull request merged**